mne/decoding/time_gen.py

```diff
@@ -408,86 +460,84 @@ def _predict_time_loop(X, estimators, cv, slices, predict_mode,
-            start += n_test_epochs
-        X = X[all_test]  # XXX JRK: Still 12 % of cpu time.
-
-        # Check that training cv and predicting cv match
-        if (len(estimators) != len(cv)) or (cv.n != len(X)):
-            raise ValueError(
-                'When `predict_mode = "cross-validation"`, the training '
-                'and predicting cv schemes must be identical.')
-    elif predict_mode != 'mean-prediction':
-        raise ValueError('`predict_mode` must be a str, "mean-prediction" or'
-                         '"cross-validation"')
     n_epochs, _, n_times = X.shape
 
-    # Checks whether the GAT is based on contiguous window of lengths = 1
-    # time-sample, ranging across the entire times. In this case, we will
-    # be able to vectorize the testing times samples.
-    expected_start = np.arange(n_times)
-    is_single_time_sample = np.array_equal([ii for sl in slices for ii in sl],
-                                           expected_start)
-    if is_single_time_sample:
-        # In simple mode, we avoid iterating over time slices.
-        slices = [slice(expected_start[0], expected_start[-1] + 1, 1)]
-    elif _warn_once.get('vectorization', True):
-        logger.warning('not vectorizing predictions across testing times, '
-                       'using a time window with length > 1')
-        _warn_once['vectorization'] = False
-    # Iterate over testing times. If is_single_time_sample, then 1 iteration.
-    y_pred = list()
-    for t, indices in enumerate(slices):
-        # Vectoring chan_times features in case of multiple time samples given
-        # to the estimators.
-        if not is_single_time_sample:
-            X_pred = X[:, :, indices].reshape(n_epochs, -1)
+    def initialize_y_pred(y_dim):
+        # y_pred can only be initialized after the first prediction, because we
+        # can't know whether it is a a categorical output or a set of
+        # probabilistic estimates.
+        # If all train time points have the same number of testing time
+        # points, then y_pred is a matrix. Else it is an array of arrays.
+        n_train = len(estimators)
+        n_test = [len(test_time_slices) for test_time_slices in train_t_slices]
+        if len(set(n_test)) == 1:
+            y_pred = np.empty((n_train, n_test[0], n_orig_epochs, y_dim))
         else:
+            y_pred = np.array([np.empty(this_n, n_orig_epochs, y_dim)
+                               for this_n in n_test])
+        return y_pred
+
+    # Loop across training times (i.e. estimators)
+    y_pred = None
+    for train_idx, (estimator_cv, test_time_slices) in enumerate(
+            zip(estimators, train_t_slices)):
+        # Checks whether predict is based on contiguous windows of lengths = 1
+        # time-sample, ranging across the entire times. In this case, we will
+        # be able to vectorize the testing times samples.
+        expected_start = np.arange(n_times)
+        is_one_time_sample = np.array_equal(
```

Comment:
A style suggestion. I see this a lot with people who have switched from Matlab (because that's kind of the convention there) -- long variable names with three or four underscores. Python let's you separate namespaces. There should be a way to come up with better variable names